### PR TITLE
I think Zimmerman subtraction works naively.

### PR DESCRIPTION
**Pull request by spaceships**
Nothing fancy, just replaced addition with subtraction and added SUB gates to the circuits too. Seems to work with circuits/sub.acirc

### CODE REVIEW

**Comment by spaceships** [comment]
Nope - this is broken. It sometimes works and sometimes doesn't. Working on it.

**Comment by amaloz** [comment]
Yes, now it's coming back to me.  I think I had tried the naive approach as well but there were some settings where this was failing.  Not sure what the correct fix is -- I think I had tried mapping subtraction to addition of the (additive) inverse.  That should work, but I don't recall why I didn't follow through with that approach.

**Comment by spaceships** [comment]
I found a bug with direct multilinear map subtraction, but I don't really understand it. About half time time, is_zero(x - x) would return false. I narrowed it down to the zero testing function.

In Tancrede's original CLT13 implementation, in the zero testing function he uses a modNear function. So he does modNear(x * pzt, q). The modNear function is defined as follows:

modNear(x,q):
    z = x mod q
    if x > q/2:
        z = z - q
    return z

Anyway, I added this to the zero tester in clt_mlm.cpp and wrote some tests. All the obfuscation tests aren't affected.

**Comment by spaceships** [comment]
Okay. That's fixed.

File: code/src/clt_mlm.cpp [hunk 1]
```diff
@@ -65,7 +65,7 @@ clt_mlm_setup(struct clt_mlm_state *s, const char *dir, const long *pows,
     s->rho = s->secparam;
     rho_f = kappa * (s->rho + alpha + 2);
     eta = rho_f + alpha + 2 * beta + s->secparam + 8;
-    nu = eta - beta - rho_f - s->secparam + 3;
+    s->nu = eta - beta - rho_f - s->secparam - 3;
```
Comment (amaloz) [inline code/src/clt_mlm.cpp]: Why -3 instead of +3?  Was I misreading the formula for nu all along?

**Comment by amaloz** [comment]
Cool!  Thanks!  Will merge soon.  First, a couple questions about the pull request:

1. Any particular reason for changing pow from long -> int?
2. Also for adding nu to the CLT MLM state?
3. See inline comment about +3 -> -3 for the setting of nu.

**Comment by spaceships** [comment]
1. Consistency with pows in clt_mlm_setup and clt_mlm_encode. One was int* and the other long*. Maybe long makes more sense? I just want them to be the same.

2. This is convenience. The function that calculates it is in mlm_setup, and I didn't want to have to recalculate it outside.

3. Yeah that's what it is on p9. I misread it too.

**Comment by spaceships** [comment]
My next pull request won't have all the extra goodies like this ;)

**Comment by amaloz** [comment]
No problem!  Merged to master.

**Pull request closed without merging**